Issue#33 types

minicole opened the pull request:
Issue #33 

Gives functionality to type effectivenesses.

![device-2015-09-01-204642](https://cloud.githubusercontent.com/assets/6855705/9620641/9b447678-50ea-11e5-8da4-28a7d2a44f20.png)

minicole committed:
Issue #33 - Type Effectiveness

Added methods so that type effectiveness is correctly shown in detailedPokemon.

Currently in this commit, there is no difference if there is a super strong 4X or normal 2X  weakness, or strong 1/4X or normal 1/2X resists to.

minicole committed:
Issue #33 - Type Effectiveness

Fixed DetailedPokemonActivity so that <1X is Strength and >1X is Weaknesses.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/DetailedPokemonActivity.java:
```diff
@@ -196,6 +197,8 @@ public void preventFurtherExecution(){
     private TextView evolutionTab;
     private TextView alternatesTab;
 
+    private final int MAX_NUMBER_OF_TYPES = 17;
```
This variable shouldn't be defined in a UI class and I also don't see it being used. It should be deleted.

minicole replied:
Forgot to remove. Done.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/DetailedPokemonActivity.java:
```diff
@@ -451,21 +454,94 @@ public void onClick(View view) {
+            Type type = types.get(0);
+            for (int i = 1; i < Type.getNumberOfTypes() + 1; i++) {
+                double effectiveDefence = type.getDefendingEffectivenessAgainst(allTypes[i]);
+                double effectiveOffence = type.getAttackingEffectivenessAgainst(allTypes[i]);
```
"effectiveOffence" isn't being used anywhere, consider deleting it.

minicole replied:
Done.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/DetailedPokemonActivity.java:
```diff
@@ -451,21 +454,94 @@ public void onClick(View view) {
+                double effectiveOffence = type.getAttackingEffectivenessAgainst(allTypes[i]);
+
+                // add the types to the correct lists
+                if (effectiveDefence == 0)
```
Have you considered changing this into a "switch" statement?

minicole replied:
I have. Unfortunately, switch does not work on doubles, and we have values of 0.25 and 0.5.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1107,4 +1116,47 @@ public boolean isDetailedNationalIDBuiltAndReady(int nationalID){
+     */
+    private double[][] buildTypeEffectivenessTable() {
+        int maxTypeID = getMaxTypeID();
+        double[][] typeEffectiveness = new double[maxTypeID][maxTypeID];
```
This local variable has the same name as the class variable. It might be best to initialize it within this method rather than returning the sparse matrix.

minicole replied:
The point of doing it this way was to make the class's 2D array a final variable. What do you think about this?

foxtrotzulu94 replied:
Since it's very unlikely to be directly accessed or modified by a class other than PokemonFactory you could just initialize within the method, similar to the "type" Type[] variable

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1107,4 +1116,47 @@ public boolean isDetailedNationalIDBuiltAndReady(int nationalID){
+
+        for (int i = 0; i < maxTypeID; i++) {
+            for (int j = 0; j < maxTypeID; j++) {
+                String[] selectionArg = {String.valueOf(i + 1), String.valueOf(j + 1)};
```
Why not initialize "i" and "j" to 1 and change the condition to "<="?

minicole replied:
Done.

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/databaseAccess/PokemonFactory.java:
```diff
@@ -1107,4 +1116,47 @@ public boolean isDetailedNationalIDBuiltAndReady(int nationalID){
+        for (int i = 0; i < maxTypeID; i++) {
+            for (int j = 0; j < maxTypeID; j++) {
+                String[] selectionArg = {String.valueOf(i + 1), String.valueOf(j + 1)};
+                Cursor cursor = database.query(TYPE_EFFECTIVENESS_TABLE, null, FROM_TYPE_ID + "=?"
```
This is O(n^2) time due to double for-loops above. Consider restructuring this query so that you load the entire table into memory and then format it into the Sparse Array.

minicole replied:
Done. Significantly faster :)

foxtrotzulu94 commented on app/src/main/java/me/quadphase/qpdex/pokemon/Type.java:
```diff
@@ -54,10 +56,10 @@ public String getName() {
-//    }
+    public double getAttackingEffectivenessAgainst (Type defendingType) {
+        PokemonFactory pokemonFactory = PokemonFactory.getPokemonFactory(null);
+        return pokemonFactory.getTypeEffectivenessTable()[this.typeID - 1][defendingType.getTypeID() - 1];
```
Careful here. You can make the sparse array directly index the types rather than having n-1. That will also avoid any sort of exceptional behaviour.

minicole replied:
Done.

foxtrot94 commented:
The application builds but crashes when opening Pokemon 0, "MissingNo."
"*Caused by: java.lang.ArrayIndexOutOfBoundsException: length=18; index=-1*"

minicole committed:
Issue#33-Type Effectiveness Table

-Fixed issues with MissingNo.
-Optimized to only go into the table once: O(n^2)->O(n).
-Removed unused variables.

foxtrot94 commented:
Just built and it's looking good. There's some optimizations you could do but they're certainly less impacting than the ones I mentioned.
For now, this should close #33.
